projects/pgai/pgai/vectorizer/embedders/litellm.py

```diff
@@ -0,0 +1,132 @@
+from collections.abc import Sequence
+from functools import cached_property
+from typing import Any, Literal
+
+import litellm
+from litellm import EmbeddingResponse as LiteLLMEmbeddingResponse  # type: ignore
+from litellm import InMemoryCache  # type: ignore
+from pydantic import BaseModel
+from typing_extensions import override
+
+from ..embeddings import (
+    BatchApiCaller,
+    Embedder,
+    EmbeddingResponse,
+    EmbeddingVector,
+    StringDocument,
+    Usage,
+    logger,
+)
+
+
+# TODO: remove this when this issue is fixed upstream: https://github.com/BerriAI/litellm/issues/7667
+# Note: we did consider building an event-loop aware in-memory cache, but the
+# additional complexity doesn't seem to be worth it.
+class NoopCache(InMemoryCache):
+    """
+    A no-op cache
+
+    This class exists because litellm's internals cause http clients to be
+    re-used across different event loops. The httpx client does not like this,
+    which causes exceptions to be thrown. Note: Not all http clients throw
+    exceptions, so we are being overly cautious with this approach.
+    """
+
+    def __init__(self):
+        super().__init__()
+
+    @override
+    def get_cache(self, key: Any, **kwargs: Any):
+        return None
+
+    @override
+    def set_cache(self, key: Any, value: Any, **kwargs: Any):
+        pass
+
+
+litellm.in_memory_llm_clients_cache = NoopCache()
+
+
+class UnknownProviderError(Exception):
+    pass
+
+
+class LiteLLM(BaseModel, Embedder):
+    """
+    Embedder that uses LiteLLM to embed documents into vector representations.
+
+    Attributes:
+        implementation (Literal["litellm"]): The literal identifier for this
+            implementation.
+        model (str): The name of the embedding model.
+        api_key_name (str): The API key name.
+        extra_options (dict): Additional litellm-specific options
+    """
+
+    implementation: Literal["litellm"]
+    model: str
+    api_key_name: str | None = None
+    extra_options: dict[str, Any] = {}
+
+    @override
+    async def embed(self, documents: list[str]) -> Sequence[EmbeddingVector]:
+        """
+        Embeds a list of documents into vectors using LiteLLM.
+
+        Args:
+            documents (list[str]): A list of documents to be embedded.
+
+        Returns:
+            Sequence[EmbeddingVector | ChunkEmbeddingError]: The embeddings or
+            errors for each document.
+        """
+        await logger.adebug(f"Chunks produced: {len(documents)}")
+        return await self._batcher.batch_chunks_and_embed(documents)
+
+    @cached_property
+    def _batcher(self) -> BatchApiCaller[StringDocument]:
+        return BatchApiCaller(self._max_chunks_per_batch(), self.call_embed_api)
+
+    @override
+    def _max_chunks_per_batch(self) -> int:
+        print(f"model: {self.model}")
```

Comment:
Might want to make this a debug log or just remove it?

Reply:
Oof - I did not intend to add that. Thanks.